ultralytics/nn/modules/head.py

```diff
@@ -80,6 +79,101 @@ def bias_init(self):
             b[-1].bias.data[:m.nc] = math.log(5 / m.nc / (640 / s) ** 2)  # cls (.01 objects, 80 classes, 640 img)
 
 
+class DetectContrastive(nn.Module):
+    """YOLOv8 Detect head for detection models using Contrastive Learning."""
+    dynamic = False  # force grid reconstruction
+    export = False  # export mode
+    shape = None
+    anchors = torch.empty(0)  # init
+    strides = torch.empty(0)  # init
+
+    def __init__(self, nc=80, ch=()):  # ch is a list of number of channels in input layer (e.g., [ch_small, ch_medium, ch_large])
+        """Initializes the YOLOv8 detection layer with specified number of classes and channels.
+        This implementation has an additional contrastive learning task for embeddings of predicted bboxes.
+        This expands (doubles) the out_channels of the last convolution layer before classification layer 
+        to split embedding dimension for classification and contrastive learning tasks.
+        """
+        super().__init__()
+        self.nc = nc  # number of classes
+        self.nl = len(ch)  # number of detection layers (e.g., small, medium, large for 3-scale detections)
+        self.reg_max = 16  # DFL channels (ch[0] // 16 to scale 4/8/12/16/20 for n/s/m/l/x)
+        c2, c3 = max((16, ch[0] // 4, self.reg_max * 4)), max(ch[0], min(self.nc, 100))  # channels
+        self.no = nc + self.reg_max * 4 + c3  # number of outputs per anchor (nc is num classes, reg_max is # predicted bboxes & 4 is bbox attributes (x, y, w, h), c3 is num features in final classification layer's embeddings)
+        self.stride = torch.zeros(self.nl)  # strides computed during build
+
+        # cv2 is for predicting bboxes (last layer's out_channel is 4 * reg_max)
+        self.cv2 = nn.ModuleList(
+            nn.Sequential(Conv(x, c2, 3), Conv(c2, c2, 3), nn.Conv2d(c2, 4 * self.reg_max, 1)) for x in ch)
+        
+        # expand out_channels for conv layer before classification layer to split embedding dimension for classification and contrastive learning tasks
+        self.c3_cls_contrastive = c3 * 2  # e.g., 64 + 64 = 128
```

Comment:
It would be clearer and more flexible if `self.c3_cls_contrastive` represented the number of channels for the embeddings, rather than the sum of classification feature vector plus embedding. For instance, right now you have no easy way to express the case where the number of channels for these two purposes is different (if you wanted to). So I would set
```
self.c3_cls_contrastive = c3
```
and later use
```
Conv(c3, self.c3 + self.c3_cls_contrastive, 3)
```
Then in `forward()` you can write more simply:
```
            embeddings_cls = self.embedding_layers[i](x[i])[:, :self.c3, :, :]  # (B, first half of n_features, H, W)
            embeddings_contrastive = self.embedding_layers[i](x[i])[:, self.c3:, :, :]  # (B, second half of n_features, H, W)
```